### PR TITLE
[Tokenizer] Add an option to specify tokenizer

### LINKED ISSUE #111: Support custom tokenizer
We should provide a clean abstraction and interface so that users can use their custom tokenizer very easily.

### LINKED ISSUE #246: Why vllm does not support Chinese input
There is a decode error in Chinese input，is token.decode

### LINKED ISSUE #259: How to mannually Set use_fast for tokenizer to False？

### LINKED ISSUE #270: The hf-internal-testing/llama-tokenizer do not support Chinese prompt
I wonder whether the vllm support Chinese or other language, because I can successfully inference with English prompt, but when I use Chinese prompt, exception raised:
```
INFO 06-27 11:11:16 tokenizer_utils.py:30] Using the LLaMA fast tokenizer in 'hf-internal-testing/llama-tokenizer' to avoid potential protobuf errors.
INFO 06-27 11:13:57 llm_engine.py:128] # GPU blocks: 247, # CPU blocks: 327
Processed prompts:   0%|          | 0/1 [00:00<?, ?it/s]Traceback (most recent call last):
  File "/mnt/lustre/sunyuhan/./scripts/testvllm.py", line 38, in <module>
    outputs = llm.generate(prompts, sampling_params)
  File "/mnt/cache/sunyuhan/miniconda3/lib/python3.10/site-packages/vllm/entrypoints/llm.py", line 114, in generate
    return self._run_engine(use_tqdm)
  File "/mnt/cache/sunyuhan/miniconda3/lib/python3.10/site-packages/vllm/entrypoints/llm.py", line 134, in _run_engine
    step_outputs = self.llm_engine.step()
  File "/mnt/cache/sunyuhan/miniconda3/lib/python3.10/site-packages/vllm/engine/llm_engine.py", line 242, in step
    self._decode_sequences(seq_groups)
  File "/mnt/cache/sunyuhan/miniconda3/lib/python3.10/site-packages/vllm/engine/llm_engine.py", line 259, in _decode_sequences
    new_token, new_output_text = detokenize_incrementally(
  File "/mnt/cache/sunyuhan/miniconda3/lib/python3.10/site-packages/vllm/engine/tokenizer_utils.py", line 68, in detokenize_incrementally
    output_text = tokenizer.convert_tokens_to_string(output_tokens)
  File "/mnt/cache/sunyuhan/miniconda3/lib/python3.10/site-packages/transformers/tokenization_utils_fast.py", line 536, in convert_tokens_to_string
    return self.backend_tokenizer.decoder.decode(tokens)
TypeError: argument 'tokens': 'NoneType' object cannot be converted to 'PyString'
```
And I also want to know how to solve this problem.

### PR DESCRIPTION
**Pull request by WoosukKwon**
Fixes #111 #246 #259 #270 #281

This PR adds `tokenizer` to the input/cli arguments. If it is `None`, vLLM uses the model name/path as the tokenizer name/path. In addition, from this PR, vLLM does not use `hf-internal-testing/llama-tokenizer` as the default tokenizer for llama models.

### CODE REVIEW

**Comment by sleepcoo** [comment]
This pr is very useful, my local test is always hard code tokenizer path

File: vllm/transformers_utils/tokenizer.py [hunk 1]
```diff
@@ -1,46 +1,44 @@
...
-        logger.info(
-            f"Using the slow tokenizer for {config.model_type} due to bugs in "
-            "the fast tokenizer. This could potentially lead to performance "
-            "degradation.")
-        kwargs["use_fast"] = False
-    return AutoTokenizer.from_pretrained(model_name, *args, **kwargs)
+            "For some LLaMA-based models, initializing the fast tokenizer may "
+            "take a long time. To eliminate the initialization time, consider "
+            f"using '{_FAST_LLAMA_TOKENIZER}' instead of the original "
+            "tokenizer.")
```
Comment (zhuohan123) [inline vllm/transformers_utils/tokenizer.py:25]: After this PR, do we need to manually specify llama to use the fast tokenizer for benchmarking?

**Review by zhuohan123 — APPROVED** [review]
LGTM! Thanks for the great work!

File: vllm/transformers_utils/tokenizer.py (hunk 1 above)
Comment (WoosukKwon) [reply vllm/transformers_utils/tokenizer.py:25]: It depends. Actually, LLaMA fast tokenizers in `lmsys/vicuna-7b-v1.3` or `huggyllama/llama-7b` work in my docker environment. So `hf-internal-testing/llama-tokenizer` is not needed when I use vLLM in my docker environment.

**Comment by 929359291** [comment]
wow is cool，bro you are so cool

**Comment by sunyuhan19981208** [comment]
THANKS VERY MUCH!

**Pull request merged**